File: app/src/main/java/com/example/kaola/myapplication/designpattern/decorator/java/factory/DumplingsFactory.java
package com.example.kaola.myapplication.designpattern.decorator.java.factory;

/**
 * @author zhangchao on 2019-07-18.
 */

public class DumplingsFactory {
    ADumplings aDumplings;

    public Dumplings getFruitDumplings() {
        aDumplings = new FruitDumplings();
        aDumplings.makeDumplings();
        return aDumplings.getDumpling();
    }

    public Dumplings getMeatDumplings() {
        aDumplings = new MeatDumplings();
        aDumplings.makeDumplings();
        return aDumplings.getDumpling();
    }
}

File: app/src/main/java/com/example/kaola/myapplication/designpattern/decorator/kotlin/factory/DumplingsFactory.kt
package com.example.kaola.myapplication.designpattern.decorator.kotlin.factory

/**
 * @author zhangchao on 2019-07-18.
 */

class DumplingsFactory {
    var mADumplings: ADumplings? = null

    fun getFruitDumplings(): Dumplings? {
        mADumplings = FruitDumplings()
        return mADumplings?.run {
            makeDumplings()
            getDumplings()
        }
    }

    fun getMeatDumplings(): Dumplings? {
        mADumplings = MeatDumplings()
        return mADumplings?.run {
            makeDumplings()
            getDumplings()
        }
    }
}